apps/lib/http_server.c

```diff
@@ -0,0 +1,399 @@
+            goto out;
+        }
+        url++;
+
+        /* Splice off the HTTP version identifier. */
+        for (end = url; *end != '\0'; end++)
+            if (*end == ' ')
+                break;
+        if (strncmp(end, " HTTP/1.", 7) != 0) {
+            log_message(prog, LOG_INFO,
+                        "Invalid GET -- bad HTTP/version string: %s", end + 1);
+            goto out;
+        }
+        *end = '\0';
+
+        /*-
+         * Skip "GET / HTTP..." requests often used by load-balancers.
+         * 'url' was incremented above to point to the first byte *after*
+         * the leading slash, so in case 'GET / ' it is now an empty string.
+         */
+        if (url[0] == '\0')
+            goto out;
+
+        len = urldecode(url);
+        if (len <= 0) {
+            log_message(prog, LOG_INFO,
+                        "Invalid GET request -- bad URL encoding: %s", url);
+            goto out;
+        }
+        if ((getbio = BIO_new_mem_buf(url, len)) == NULL
+            || (b64 = BIO_new(BIO_f_base64())) == NULL) {
+            log_message(prog, LOG_ERR,
+                        "Could not allocate base64 bio with size = %d", len);
+            BIO_free_all(cbio);
+            *pcbio = NULL;
+            ret = -1;
+            goto out;
+        }
+        BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);
+        getbio = BIO_push(b64, getbio);
+    } else if (strncmp(reqbuf, "POST ", 5) != 0) {
+        log_message(prog, LOG_INFO,
+                    "HTTP request does not start with GET/POST: %s", reqbuf);
+        /* TODO provide better diagnosis in case client tries TLS */
+        goto out;
+    }
+
+    /* Read and skip past the headers. */
+    for (;;) {
+        len = BIO_gets(cbio, inbuf, sizeof(inbuf));
+        if (len <= 0) {
+            log_message(prog, LOG_ERR,
+                        "Error skipping remaining HTTP headers");
+            goto out;
+        }
+        if ((inbuf[0] == '\r') || (inbuf[0] == '\n'))
+            break;
+    }
+
+# ifdef HTTP_DAEMON
+    /* Clear alarm before we close the client socket */
+    alarm(0);
+    timeout = 0;
+# endif
+
+    /* Try to read and parse request */
+    req = ASN1_item_d2i_bio(it, getbio != NULL ? getbio : cbio, NULL);
+    if (req == NULL)
+        log_message(prog, LOG_ERR, "Error parsing request");
+
+    *preq = req;
+
+ out:
+    BIO_free_all(getbio);
+# ifdef HTTP_DAEMON
+    if (timeout > 0)
+        alarm(0);
+    acfd = (int)INVALID_SOCKET;
+# endif
+    return ret;
+}
+
+int http_server_send_resp(BIO *cbio, const char *content_type,
+                          const ASN1_ITEM *it, const ASN1_VALUE *val)
+{
+    int ret = BIO_printf(cbio, "HTTP/1.0 200 OK\r\nContent-type: %s\r\n"
+                         "Content-Length: %d\r\n\r\n", content_type,
+                         ASN1_item_i2d(val, NULL, it)) > 0
+            && ASN1_item_i2d_bio(it, cbio, val) > 0;
```

Comment:
There's an assumption here that `cbio` is not performing a non-trivial encoding that changes the output length.  That should perhaps be noted in comments above the function.

Reply:
I've added a respective comment/note to the implementation and in the header file.